external-dns crashlooping when using godaddy provider - panic: invalid argument to Int63n

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

-->

**What happened**:

Start around 11:55 EST today (Monday, Nov 11th, 2024) my external DNS deployment started crashlooping out of the blue. No configuration change has been made in the last 2 weeks. 

As far as I can tell, it seems to be failing because the `Retry-After` header is either missing or its value can't be parsed to Int64 here: https://github.com/kubernetes-sigs/external-dns/blob/ea6d44b0d81c635ee275b51ab777a17151a08e9d/provider/godaddy/client.go#L233

In addition, what is strange is that the 429 response, which is what would trigger the above code, only seems to happen in the external-dns controller. Calling the GoDaddy API with the same API Key/Secret from the same IP does not trigger a 429 response, yet the external-dns pod is consistently unable to start.

**What you expected to happen**:

External-dns should not crash. At best, there should be a more graceful handling of a missing or invalid `Retry-After` header. Maybe defaulting to 60 seconds?

**How to reproduce it (as minimally and precisely as possible)**:

Still unclear what the root cause is. Will setup a local cluster and report back if I find anything.

**Anything else we need to know?**:

Logs: 

```
panic: invalid argument to Int63n

goroutine 1 [running]:
math/rand.(*Rand).Int63n(0x3bbc4c9?, 0xb?)
    math/rand/rand.go:122 +0xcb
math/rand.Int63n(0x0)
    math/rand/rand.go:443 +0x25
sigs.k8s.io/external-dns/provider/godaddy.(*Client).Do(0xc000ee03c0, 0xc0016fadc0)
    sigs.k8s.io/external-dns/provider/godaddy/client.go:235 +0x127
sigs.k8s.io/external-dns/provider/godaddy.(*Client).CallAPIWithContext(0xc000ee03c0, {0x4261490, 0x6574a40}, {0x3ba7c15?, 0xc000ecd550?}, {0x3c45c1f?, 0x30?}, {0x0?, 0x0?}, {0x0, ...}, ...)
    sigs.k8s.io/external-dns/provider/godaddy/client.go:301 +0x127
sigs.k8s.io/external-dns/provider/godaddy.(*Client).CallAPI(...)
    sigs.k8s.io/external-dns/provider/godaddy/client.go:272
sigs.k8s.io/external-dns/provider/godaddy.(*Client).Get(...)
    sigs.k8s.io/external-dns/provider/godaddy/client.go:143
sigs.k8s.io/external-dns/provider/godaddy.(*Client).validate(0x47305f?)
    sigs.k8s.io/external-dns/provider/godaddy/client.go:342 +0x4b
sigs.k8s.io/external-dns/provider/godaddy.NewClient(0x1?, {0x7ffef2140a54, 0x23}, {0x7ffef2140a8d, 0x16})
    sigs.k8s.io/external-dns/provider/godaddy/client.go:131 +0x198
sigs.k8s.io/external-dns/provider/godaddy.NewGoDaddyProvider({0x0?, 0xc03ca9?}, {{0xc000e8cd70, 0x1, 0x1}, {0x0, 0x0, 0x0}, 0x0, 0x0}, ...)
    sigs.k8s.io/external-dns/provider/godaddy/godaddy.go:142 +0x46
main.main()
    sigs.k8s.io/external-dns/main.go:342 +0x206f
Stream closed EOF for external-dns/external-dns-godaddy-cmls-75d996f88f-wp8gx (external-dns)

```

**Environment**:
- External-DNS version (use `external-dns --version`): 1.15.0 (also happening in 1.14.5)
- DNS provider: godaddy
- Others:

--- a/provider/godaddy/client.go
+++ b/provider/godaddy/client.go
@@ -152,56 +152,56 @@ func NewClient(useOTE bool, apiKey, apiSecret string) (*Client, error) {
 
 // Get is a wrapper for the GET method
 func (c *Client) Get(url string, resType interface{}) error {
-	return c.CallAPI("GET", url, nil, resType, true)
+	return c.CallAPI("GET", url, nil, resType)
 }
 
 // Patch is a wrapper for the PATCH method
 func (c *Client) Patch(url string, reqBody, resType interface{}) error {
-	return c.CallAPI("PATCH", url, reqBody, resType, true)
+	return c.CallAPI("PATCH", url, reqBody, resType)
 }
 
 // Post is a wrapper for the POST method
 func (c *Client) Post(url string, reqBody, resType interface{}) error {
-	return c.CallAPI("POST", url, reqBody, resType, true)
+	return c.CallAPI("POST", url, reqBody, resType)
 }
 
 // Put is a wrapper for the PUT method
 func (c *Client) Put(url string, reqBody, resType interface{}) error {
-	return c.CallAPI("PUT", url, reqBody, resType, true)
+	return c.CallAPI("PUT", url, reqBody, resType)
 }
 
 // Delete is a wrapper for the DELETE method
 func (c *Client) Delete(url string, resType interface{}) error {
-	return c.CallAPI("DELETE", url, nil, resType, true)
+	return c.CallAPI("DELETE", url, nil, resType)
 }
 
 // GetWithContext is a wrapper for the GET method
 func (c *Client) GetWithContext(ctx context.Context, url string, resType interface{}) error {
-	return c.CallAPIWithContext(ctx, "GET", url, nil, resType, true)
+	return c.CallAPIWithContext(ctx, "GET", url, nil, resType)
 }
 
 // PatchWithContext is a wrapper for the PATCH method
 func (c *Client) PatchWithContext(ctx context.Context, url string, reqBody, resType interface{}) error {
-	return c.CallAPIWithContext(ctx, "PATCH", url, reqBody, resType, true)
+	return c.CallAPIWithContext(ctx, "PATCH", url, reqBody, resType)
 }
 
 // PostWithContext is a wrapper for the POST method
 func (c *Client) PostWithContext(ctx context.Context, url string, reqBody, resType interface{}) error {
-	return c.CallAPIWithContext(ctx, "POST", url, reqBody, resType, true)
+	return c.CallAPIWithContext(ctx, "POST", url, reqBody, resType)
 }
 
 // PutWithContext is a wrapper for the PUT method
 func (c *Client) PutWithContext(ctx context.Context, url string, reqBody, resType interface{}) error {
-	return c.CallAPIWithContext(ctx, "PUT", url, reqBody, resType, true)
+	return c.CallAPIWithContext(ctx, "PUT", url, reqBody, resType)
 }
 
 // DeleteWithContext is a wrapper for the DELETE method
 func (c *Client) DeleteWithContext(ctx context.Context, url string, resType interface{}) error {
-	return c.CallAPIWithContext(ctx, "DELETE", url, nil, resType, true)
+	return c.CallAPIWithContext(ctx, "DELETE", url, nil, resType)
 }
 
 // NewRequest returns a new HTTP request
-func (c *Client) NewRequest(method, path string, reqBody interface{}, needAuth bool) (*http.Request, error) {
+func (c *Client) NewRequest(method, path string, reqBody interface{}) (*http.Request, error) {
 	var body []byte
 	var err error
 
@@ -240,8 +240,11 @@ func (c *Client) Do(req *http.Request) (*http.Response, error) {
 
 	c.Ratelimiter.Wait(req.Context())
 	resp, err := c.Client.Do(req)
+	if err != nil {
+		return nil, err
+	}
 	// In case of several clients behind NAT we still can hit rate limit
-	for i := 1; i < 3 && err == nil && resp.StatusCode == 429; i++ {
+	for i := 1; i < 3 && resp != nil && resp.StatusCode == 429; i++ {
 		retryAfter, err := strconv.ParseInt(resp.Header.Get("Retry-After"), 10, 0)
 		if err != nil {
 			log.Error("Rate-limited response did not contain a valid Retry-After header, quota likely exceeded")
@@ -256,9 +259,6 @@ func (c *Client) Do(req *http.Request) (*http.Response, error) {
 
 		c.Ratelimiter.Wait(req.Context())
 		resp, err = c.Client.Do(req)
-	}
-	if err != nil {
-		return nil, err
 	}
 	if c.Logger != nil {
 		c.Logger.LogResponse(resp)
@@ -284,8 +284,8 @@ func (c *Client) Do(req *http.Request) (*http.Response, error) {
 //
 // If everything went fine, unmarshall response into resType and return nil
 // otherwise, return the error
-func (c *Client) CallAPI(method, path string, reqBody, resType interface{}, needAuth bool) error {
-	return c.CallAPIWithContext(context.Background(), method, path, reqBody, resType, needAuth)
+func (c *Client) CallAPI(method, path string, reqBody, resType interface{}) error {
+	return c.CallAPIWithContext(context.Background(), method, path, reqBody, resType)
 }
 
 // CallAPIWithContext is the lowest level call helper. If needAuth is true,
@@ -308,8 +308,8 @@ func (c *Client) CallAPI(method, path string, reqBody, resType interface{}, need
 //
 // If everything went fine, unmarshall response into resType and return nil
 // otherwise, return the error
-func (c *Client) CallAPIWithContext(ctx context.Context, method, path string, reqBody, resType interface{}, needAuth bool) error {
-	req, err := c.NewRequest(method, path, reqBody, needAuth)
+func (c *Client) CallAPIWithContext(ctx context.Context, method, path string, reqBody, resType interface{}) error {
+	req, err := c.NewRequest(method, path, reqBody)
 	if err != nil {
 		return err
 	}

--- a/provider/godaddy/client.go
+++ b/provider/godaddy/client.go
@@ -28,6 +28,7 @@ import (
 	"strconv"
 	"time"
 
+	log "github.com/sirupsen/logrus"
 	"golang.org/x/time/rate"
 
 	"sigs.k8s.io/external-dns/pkg/apis/externaldns"
@@ -39,6 +40,11 @@ const DefaultTimeout = 180 * time.Second
 // Errors
 var (
 	ErrAPIDown = errors.New("godaddy: the GoDaddy API is down")
+)
+
+// error codes
+const (
+	ErrCodeQuotaExceeded = "QUOTA_EXCEEDED"
 )
 
 // APIError error
@@ -129,7 +135,13 @@ func NewClient(useOTE bool, apiKey, apiSecret string) (*Client, error) {
 
 	// Get and check the configuration
 	if err := client.validate(); err != nil {
-		return nil, err
+		var apiErr *APIError
+		// Quota Exceeded errors are limited to the endpoint being called. Other endpoints are not affected when we hit
+		// the quota limit on the endpoint used for validation. We can safely ignore this error.
+		// Quota limits on other endpoints will be logged by their respective calls.
+		if ok := errors.As(err, &apiErr); ok && apiErr.Code != ErrCodeQuotaExceeded {
+			return nil, err
+		}
 	}
 	return &client, nil
 }
@@ -230,7 +242,11 @@ func (c *Client) Do(req *http.Request) (*http.Response, error) {
 	resp, err := c.Client.Do(req)
 	// In case of several clients behind NAT we still can hit rate limit
 	for i := 1; i < 3 && err == nil && resp.StatusCode == 429; i++ {
-		retryAfter, _ := strconv.ParseInt(resp.Header.Get("Retry-After"), 10, 0)
+		retryAfter, err := strconv.ParseInt(resp.Header.Get("Retry-After"), 10, 0)
+		if err != nil {
+			log.Error("Rate-limited response did not contain a valid Retry-After header, quota likely exceeded")
+			break
+		}
 
 		jitter := rand.Int63n(retryAfter)
 		retryAfterSec := retryAfter + jitter/2

--- a/provider/godaddy/client.go
+++ b/provider/godaddy/client.go
@@ -259,6 +259,9 @@ func (c *Client) Do(req *http.Request) (*http.Response, error) {
 
 		c.Ratelimiter.Wait(req.Context())
 		resp, err = c.Client.Do(req)
+		if err != nil {
+			return nil, fmt.Errorf("doing request after waiting for retry after: %w", err)
+		}
 	}
 	if c.Logger != nil {
 		c.Logger.LogResponse(resp)

--- a/provider/godaddy/client_test.go
+++ b/provider/godaddy/client_test.go
@@ -1,0 +1,56 @@
+package godaddy
+
+import (
+	"errors"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"golang.org/x/time/rate"
+)
+
+// Tests that
+func TestClient_DoWhenQuotaExceeded(t *testing.T) {
+	assert := assert.New(t)
+
+	// Mock server to return 429 with a JSON payload
+	mockServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusTooManyRequests)
+		_, err := w.Write([]byte(`{"code": "QUOTA_EXCEEDED", "message": "rate limit exceeded"}`))
+		if err != nil {
+			t.Fatalf("Failed to write response: %v", err)
+		}
+	}))
+	defer mockServer.Close()
+
+	client := Client{
+		APIKey:      "",
+		APISecret:   "",
+		APIEndPoint: mockServer.URL,
+		Client:      &http.Client{},
+		// Add one token every second
+		Ratelimiter: rate.NewLimiter(rate.Every(time.Second), 60),
+		Timeout:     DefaultTimeout,
+	}
+
+	req, err := client.NewRequest("GET", "/v1/domains/example.net/records", nil, false)
+	if err != nil {
+		t.Fatalf("Failed to create request: %v", err)
+	}
+
+	resp, err := client.Do(req)
+	assert.Nil(err, "A CODE_EXCEEDED response should not return an error")
+	assert.Equal(http.StatusTooManyRequests, resp.StatusCode, "Expected a 429 response")
+
+	respContents := GDErrorResponse{}
+	err = client.UnmarshalResponse(resp, &respContents)
+	if assert.NotNil(err) {
+		var apiErr *APIError
+		errors.As(err, &apiErr)
+		assert.Equal("QUOTA_EXCEEDED", apiErr.Code)
+		assert.Equal("rate limit exceeded", apiErr.Message)
+	}
+}

--- a/provider/godaddy/client_test.go
+++ b/provider/godaddy/client_test.go
@@ -1,3 +1,19 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
 package godaddy
 
 import (

--- a/provider/godaddy/client_test.go
+++ b/provider/godaddy/client_test.go
@@ -36,7 +36,7 @@ func TestClient_DoWhenQuotaExceeded(t *testing.T) {
 		Timeout:     DefaultTimeout,
 	}
 
-	req, err := client.NewRequest("GET", "/v1/domains/example.net/records", nil, false)
+	req, err := client.NewRequest("GET", "/v1/domains/example.net/records", nil)
 	if err != nil {
 		t.Fatalf("Failed to create request: %v", err)
 	}

